src/accumulate.jl

```diff
@@ -3,122 +3,138 @@
 ## COV_EXCL_START
 
 # partial scan of individual thread blocks within a grid
-# work-efficient implementation after Blelloch (1990)
+# Parallel Prefix Scan using warp-level primitives
 #
-# number of threads needs to be a power-of-2
+# https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda
+# 
 #
 # performance TODOs:
-# - shuffle
-# - warp-aggregate atomics
-# - the ND case is quite a bit slower than the 1D case (not using Cartesian indices,
-#   before 35fcbde1f2987023229034370b0c9091e18c4137). optimize or special-case?
-function partial_scan(op::Function, output::CuDeviceArray{T}, input::CuDeviceArray,
-                      Rdim, Rpre, Rpost, Rother, neutral, init,
-                      ::Val{inclusive}=Val(true)) where {T, inclusive}
+# - Multiple elements per thread 
+# - Use warpSize instead of hardcoding and move to GPUArrays
+function partial_scan(op::Function, output::CuDeviceArray{T}, input, aggregates,
+                        Rdim, Rpre, Rpost, Rother, neutral) where {T}
     threads = blockDim().x
     thread = threadIdx().x
-    block = blockIdx().x
+    warps = threads ÷ 32
+    mask = typemax(UInt32)
 
-    temp = @cuDynamicSharedMem(T, (2*threads,))
+    temp = @cuDynamicSharedMem(T, (threads,))
 
     # iterate the main dimension using threads and the first block dimension
-    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
```

Comment:
I found the variable names `tid`, `wid`, `bid`, `wid` more pneumonic to their purpose here rather than `i` and `j`. We can change other places in the package to keep this uniform. Your call obviously but I think its better my way and makes the code easier to understand.

Reply:
`tid` and `bid` make me think of thread ID and block ID, which it doesn't mean here, right? Both are an index across the entire iteration space. So I've reverted that for now to keep the diff smaller and make it easier to review.